client.go

```diff
@@ -87,11 +240,252 @@ func (c *Client) SendSTUNRequest(serverIP net.IP, serverPort int) (net.Addr, err
+		stun.Fingerprint,
+	)
+	if err != nil {
+		return nil, err
+	}
+	trRes, err := c.PerformTransaction(msg, c.turnServ, false)
+	if err != nil {
+		return nil, err
+	}
+
+	res := trRes.Msg
+
+	// Anonymous allocate failed, trying to authenticate.
+	var nonce stun.Nonce
+	if err := nonce.GetFrom(res); err != nil {
+		return nil, err
+	}
+	if err := c.realm.GetFrom(res); err != nil {
+		return nil, err
+	}
+	c.realm = append([]byte(nil), c.realm...)
+	c.integrity = stun.NewLongTermIntegrity(
+		c.username.String(), c.realm.String(), c.password,
+	)
+	// Trying to authorize.
+	msg, err = stun.Build(
+		stun.TransactionID,
+		stun.NewType(stun.MethodAllocate, stun.ClassRequest),
+		turn.RequestedTransportUDP,
+		&c.username,
+		&c.realm,
+		&nonce,
+		&c.integrity,
+		stun.Fingerprint,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	trRes, err = c.PerformTransaction(msg, c.turnServ, false)
+	if err != nil {
+		return nil, err
+	}
+	res = trRes.Msg
+
+	if res.Type.Class == stun.ClassErrorResponse {
+		var code stun.ErrorCodeAttribute
+		if err = code.GetFrom(res); err == nil {
+		}
+		return nil, fmt.Errorf("%s (error %s)", res.Type, code)
+	}
+
+	// Getting relayed addresses from response.
+	var relayed turn.RelayedAddress
+	if err := relayed.GetFrom(res); err != nil {
+		return nil, err
+	}
+
+	// Getting lifetime from response
+	var lifetime turn.Lifetime
+	if err := lifetime.GetFrom(res); err != nil {
+		return nil, err
+	}
+
+	relayedConn := client.NewUDPConn(&client.UDPConnConfig{
+		Observer:  c,
+		Relayed:   relayed,
+		Integrity: c.integrity,
+		Nonce:     nonce,
+		Lifetime:  lifetime.Duration,
+		Log:       c.log,
+	})
+
+	c.relayedConn = relayedConn
+
+	return relayedConn, nil
+}
+
+// PerformTransaction performs STUN transaction
+func (c *Client) PerformTransaction(msg *stun.Message, to net.Addr, dontWait bool) (client.TransactionResult, error) {
+	resCh := make(chan client.TransactionResult)
+	trKey := b64.StdEncoding.EncodeToString(msg.TransactionID[:])
+	tr := &transaction{
+		key:      trKey,
+		raw:      msg.Raw,
+		to:       to,
+		interval: c.rto,
+		resultCh: resCh,
+	}
+	c.trMap[trKey] = tr
+
+	c.log.Tracef("start %s transaction %s to %s", msg.Type, trKey, tr.to.String())
+	_, err := c.conn.WriteTo(msg.Raw, to)
+	if err != nil {
+		return client.TransactionResult{}, err
+	}
+
+	tr.startRtxTimer(c.onRtxTimeout)
+
+	// If dontWait is true, get the transaction going and return immediately
+	if dontWait {
+		return client.TransactionResult{}, nil
+	}
+
+	res := <-resCh
+	if res.Err != nil {
+		return res, res.Err
+	}
+	return res, nil
+}
+
+// HandleInbound handles data received.
+func (c *Client) HandleInbound(data []byte, from net.Addr) (bool, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	var err error
+
+	switch {
+	case stun.IsMessage(data):
+		err = c.handleSTUNMessage(data, from)
+	case turn.IsChannelData(data):
+		err = c.handleChannelData(data)
+	default:
+		return false, nil // unhandled
```

Comment:
Might be worth a trace/debug message

Reply:
Yes, I will add a debug message. (i will implement channel binding soon)

Reply:
Added!